core/store/src/db/rocksdb/snapshot.rs

```diff
@@ -0,0 +1,182 @@
+        let path: std::path::PathBuf = unsafe { core::mem::transmute(self) };
+
+        tracing::info!(target: "db", snapshot_path=%path.display(),
+                       "Deleting the database snapshot");
+        if let Err(err) = std::fs::remove_dir_all(&path) {
+            tracing::error!(target: "db", snapshot_path=%path.display(), ?err,
+                            "Failed to delete the database snapshot");
+            tracing::error!(target: "db", snapshot_path=%path.display(),
+                            "Please delete the snapshot manually before");
+        }
+    }
+}
+
+impl std::ops::Drop for Snapshot {
+    /// If the checkpoint hasn’t been deleted, log information about it.
+    ///
+    /// If the checkpoint hasn’t been deleted with [`Self::remove`] method, this
+    /// will log information about where the checkpoint resides in the file
+    /// system and how to recover data from it.
+    fn drop(&mut self) {
+        tracing::info!(target: "db", snapshot_path=%self.0.display(),
+                       "In case of issues, the database can be recovered \
+                        from the database snapshot");
+        tracing::info!(target: "db", snapshot_path=%self.0.display(),
+                       "To recover from the snapshot, delete files in the \
+                        database directory and replace them with contents \
+                        of the snapshot directory");
+    }
+}
+
+#[test]
+fn test_snapshot_creation() {
+    use assert_matches::assert_matches;
+
+    let (tmpdir, opener) = crate::Store::test_opener();
+    let path = tmpdir.path().join("cp");
+
+    // Create the database
+    core::mem::drop(opener.open());
+
+    // Creating snapshot should work now.
+    let snapshot = opener.new_migration_snapshot(path.clone()).unwrap();
+
+    // Snapshot already exists so cannot create a new one.
+    assert_matches!(
+        opener.new_migration_snapshot(path.clone()),
+        Err(SnapshotError::AlreadyExists(_))
+    );
+
+    snapshot.remove();
+
+    // This should work correctly again since the snapshot has been removed.
+    opener.new_migration_snapshot(path.clone()).unwrap();
+
+    // And this again should fail.  We don’t remove the snapshot in
+    // Snapshot::drop.
+    assert_matches!(
+        opener.new_migration_snapshot(path.clone()),
+        Err(SnapshotError::AlreadyExists(_))
+    );
+}
+
+/// Tests that reading data from a snapshot is possible.
+#[test]
+fn test_snapshot_recovery() {
+    const KEY: &[u8] = b"key";
+    const COL: crate::DBCol = crate::DBCol::BlockMisc;
+
+    let (tmpdir, opener) = crate::Store::test_opener();
+    let path = tmpdir.path().join("cp");
+
+    // Populate some data
+    {
+        let store = opener.open().unwrap();
+        let mut update = store.store_update();
+        update.set_raw_bytes(COL, KEY, b"value");
+        update.commit().unwrap();
+    }
+
+    // Create snapshot
+    std::thread::sleep(std::time::Duration::from_millis(5_000));
```

Comment:
Do we really have to wait for 5 seconds here? I don't really understand why this is here in the first place tbh.

Reply:
Sorry, leftover from testing.